src/cosim/execution_runner.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ *  This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+#include "cosim/execution_runner.hpp"
+
+#include <atomic>
+#include <future>
+
+namespace cosim
+{
+
+class execution_runner::impl
+{
+
+public:
+    explicit impl(execution& exec)
+        : exec_(exec)
+        , stopped_(true)
+        , timer_()
+    { }
+
+    ~impl() = default;
+
+    impl(const impl&) = delete;
+    impl& operator=(const impl&) = delete;
+
+    impl(impl&&) = delete;
+    impl& operator=(impl&&) = delete;
+
+    [[nodiscard]] bool is_running() const noexcept
+    {
+        return !stopped_;
+    }
+
+    void stop_simulation()
+    {
+        stopped_ = true;
+    }
+
+    std::future<bool> simulate_until(std::optional<time_point> endTime)
+    {
+        if (is_running()) {
+            throw std::logic_error("Simulation is already running!");
+        }
+
+        stopped_ = false;
+        return std::async(std::launch::async, [this, endTime]{
+            timer_.start(exec_.current_time());
+            duration stepSize;
+            do {
+                stepSize = exec_.step();
+                timer_.sleep(exec_.current_time());
+            } while (!stopped_ && !timed_out(endTime, exec_.current_time(), stepSize));
+            bool isStopped = stopped_;
+            stopped_ = true;
+            return !isStopped;
+        });
+    }
+
+    [[nodiscard]] std::shared_ptr<real_time_config> get_real_time_config() const
+    {
+        return timer_.get_real_time_config();
+    }
+
+    [[nodiscard]] std::shared_ptr<const real_time_metrics> get_real_time_metrics() const
+    {
+        return timer_.get_real_time_metrics();
+    }
+
+
+private:
+    execution& exec_;
+
+    std::atomic<bool> stopped_;
+    real_time_timer timer_;
+
+    static bool timed_out(std::optional<time_point> endTime, time_point currentTime, duration stepSize)
```

Comment:
This exact method is also implemented and used in `execution`, do we need it both places?